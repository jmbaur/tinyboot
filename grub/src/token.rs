use std::fmt::Display;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Token {
    Ampersand,
    Case,
    CloseBrace,
    CloseBracket,
    Comment(String),
    Do,
    Done,
    Elif,
    Else,
    Esac,
    Fi,
    For,
    Function,
    If,
    Illegal(char),
    In,
    Literal(String),
    Newline,
    OpenBrace,
    Pipe,
    Select,
    Semicolon,
    Then,
    Time,
    Until,
    Value(String),
    While,
}

impl Display for Token {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{self:?}")
    }
}
