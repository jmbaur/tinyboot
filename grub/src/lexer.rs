use crate::token::Token;
use std::{iter::Peekable, str::Chars};

const LINE_FEED: char = 0xa as char;

pub struct Lexer<'a> {
    current_char: char,
    src: Peekable<Chars<'a>>,
}

impl<'a> Lexer<'a> {
    pub fn new(input: &'a str) -> Self {
        Self {
            src: input.chars().peekable(),
            current_char: 0 as char,
        }
    }

    fn read_other(&mut self) -> Token {
        let mut other_str = String::from(self.current_char);
        while let Some(&char) = self.src.peek() {
            if char.is_ascii_whitespace() || char == ';' {
                break;
            }

            other_str.push(self.src.next().expect("peek is not None"));
        }

        match other_str.as_str() {
            "case" => Token::Case,
            "do" => Token::Do,
            "done" => Token::Done,
            "elif" => Token::Elif,
            "else" => Token::Else,
            "esac" => Token::Esac,
            "fi" => Token::Fi,
            "for" => Token::For,
            "function" => Token::Function,
            "if" => Token::If,
            "in" => Token::In,
            "select" => Token::Select,
            "then" => Token::Then,
            "time" => Token::Time,
            "until" => Token::Until,
            "while" => Token::While,
            _ => Token::Value(other_str),
        }
    }

    fn read_comment(&mut self) -> Token {
        let mut comment_str = String::new();

        'outer: while let Some(char) = self.src.next() {
            // Determine if this is a multi-line comment.
            'line_feed: {
                if char == LINE_FEED {
                    while let Some(next) = self.src.peek() {
                        // Multi-line comments can have all-whitespace in front of the octothorpes.
                        if next.is_ascii_whitespace() {
                            // Not a multi-line comment since there were two consecutive line feed
                            // characters.
                            if next == &LINE_FEED {
                                break 'outer;
                            }
                            _ = self.src.next().expect("peek is not None");
                            continue;
                        }

                        // Is a multi-line comment, continue building the token value.
                        if next == &'#' {
                            _ = self.src.next().expect("peek is not None");
                            break 'line_feed;
                        }

                        // Not a multi-line comment, finish the comment token.
                        break 'outer;
                    }
                }
            }

            comment_str.push(char);
        }

        Token::Comment(comment_str.trim().to_string())
    }

    fn read_quoted_val(&mut self) -> Token {
        if self.current_char != '"' {
            return Token::Illegal(self.current_char);
        }
        let mut str_val = String::new();
        for char in self.src.by_ref() {
            if char == '"' {
                break;
            }
            str_val.push(char);
        }
        Token::Value(str_val)
    }

    fn read_literal(&mut self) -> Token {
        if self.current_char != '\'' {
            return Token::Illegal(self.current_char);
        }
        let mut str_val = String::new();
        for char in self.src.by_ref() {
            if char == '\'' {
                break;
            }
            str_val.push(char);
        }
        Token::Literal(str_val)
    }
}

impl Iterator for Lexer<'_> {
    type Item = Token;

    fn next(&mut self) -> Option<Self::Item> {
        loop {
            let Some(char) = self.src.next() else { return None; };
            if char.is_ascii_whitespace() {
                if char == LINE_FEED {
                    return Some(Token::Newline);
                }
                continue;
            }
            self.current_char = char;
            return Some(match self.current_char {
                '"' => self.read_quoted_val(),
                '#' => self.read_comment(),
                '&' => Token::Ampersand,
                ';' => Token::Semicolon,
                '[' => Token::Value(String::from("[")),
                '\'' => self.read_literal(),
                ']' => Token::CloseBracket,
                '{' => Token::OpenBrace,
                '|' => Token::Pipe,
                '}' => Token::CloseBrace,
                _ => self.read_other(),
            });
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn tokenize(input: &str) -> Result<Vec<Token>, String> {
        let l = Lexer::new(input);

        let mut tokens = Vec::new();
        for token in l {
            tokens.push(token);
        }

        Ok(tokens)
    }

    #[test]
    fn whitespace() {
        assert_eq!(
            tokenize(
                r#"
                "#
            )
            .unwrap(),
            vec![Token::Newline]
        );
    }

    #[test]
    fn expressions() {
        assert_eq!(
            tokenize("string1 == string2").unwrap(),
            vec![
                Token::Value("string1".to_string()),
                Token::Value("==".to_string()),
                Token::Value("string2".to_string())
            ],
        );
        assert_eq!(
            tokenize("integer1 -gt integer2").unwrap(),
            vec![
                Token::Value("integer1".to_string()),
                Token::Value("-gt".to_string()),
                Token::Value("integer2".to_string()),
            ]
        );
    }

    #[test]
    fn set_command() {
        assert_eq!(
            tokenize("set foo=bar").unwrap(),
            vec![
                Token::Value("set".to_string()),
                Token::Value("foo=bar".to_string()),
            ]
        );
    }

    #[test]
    fn simple_expression() {
        assert_eq!(
            tokenize(
                r#"if [ $default -ne 0 ]; then
                     set default=0
                   fi"#
            )
            .unwrap(),
            vec![
                Token::If,
                Token::Value("[".to_string()),
                Token::Value("$default".to_string()),
                Token::Value("-ne".to_string()),
                Token::Value("0".to_string()),
                Token::CloseBracket,
                Token::Semicolon,
                Token::Then,
                Token::Newline,
                Token::Value("set".to_string()),
                Token::Value("default=0".to_string()),
                Token::Newline,
                Token::Fi,
            ]
        );
    }

    #[test]
    fn menuentry() {
        assert_eq!(
            tokenize("menuentry { linux /path/to/linux; }").unwrap(),
            vec![
                Token::Value("menuentry".to_string()),
                Token::OpenBrace,
                Token::Value("linux".to_string()),
                Token::Value("/path/to/linux".to_string()),
                Token::Semicolon,
                Token::CloseBrace,
            ]
        );
    }

    #[test]
    fn comment() {
        assert_eq!(
            tokenize("foo # bar").unwrap(),
            vec![
                Token::Value("foo".to_string()),
                Token::Comment("bar".to_string()),
            ]
        );

        assert_eq!(
            tokenize(
                r#"# foo
                   # bar

                   # baz"#
            )
            .unwrap(),
            vec![
                Token::Comment("foo\n bar".to_string()),
                Token::Newline,
                Token::Comment("baz".to_string()),
            ]
        );
    }

    #[test]
    fn device_syntax() {
        assert_eq!(
            tokenize("(hd0,1)").unwrap(),
            vec![Token::Value("(hd0,1)".to_string()),]
        );
    }

    #[test]
    fn full_example() {
        let tokens = tokenize(include_str!("../testdata/grub.cfg")).unwrap();
        assert!(!tokens.iter().any(|tok| matches!(tok, Token::Illegal(_))));
    }
}
