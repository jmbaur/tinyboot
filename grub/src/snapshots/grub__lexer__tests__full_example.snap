---
source: grub/src/lexer.rs
expression: tokens
---
[
    Comment(
        "Automatically generated.  DO NOT EDIT THIS FILE!",
    ),
    Newline,
    Value(
        "search",
    ),
    Value(
        "--set=drive1",
    ),
    Value(
        "--fs-uuid",
    ),
    Value(
        "BB22-99EC",
    ),
    Newline,
    If,
    Value(
        "[",
    ),
    Value(
        "-s",
    ),
    Value(
        "$prefix/grubenv",
    ),
    CloseBracket,
    Semicolon,
    Then,
    Newline,
    Value(
        "load_env",
    ),
    Newline,
    Fi,
    Newline,
    Newline,
    Comment(
        "‘grub-reboot’ sets a one-time saved entry, which we process here and\n then delete.",
    ),
    If,
    Value(
        "[",
    ),
    Value(
        "${next_entry}",
    ),
    CloseBracket,
    Semicolon,
    Then,
    Newline,
    Value(
        "set",
    ),
    Value(
        "default=\"${next_entry}\"",
    ),
    Newline,
    Value(
        "set",
    ),
    Value(
        "next_entry=",
    ),
    Newline,
    Value(
        "save_env",
    ),
    Value(
        "next_entry",
    ),
    Newline,
    Value(
        "set",
    ),
    Value(
        "timeout=1",
    ),
    Newline,
    Value(
        "set",
    ),
    Value(
        "boot_once=true",
    ),
    Newline,
    Else,
    Newline,
    Value(
        "set",
    ),
    Value(
        "default=0",
    ),
    Newline,
    Value(
        "set",
    ),
    Value(
        "timeout=5",
    ),
    Newline,
    Fi,
    Newline,
    Newline,
    Function,
    Value(
        "savedefault",
    ),
    OpenBrace,
    Newline,
    If,
    Value(
        "[",
    ),
    Value(
        "-z",
    ),
    Value(
        "${boot_once}",
    ),
    CloseBracket,
    Semicolon,
    Then,
    Newline,
    Value(
        "saved_entry=\"${chosen}\"",
    ),
    Newline,
    Value(
        "save_env",
    ),
    Value(
        "saved_entry",
    ),
    Newline,
    Fi,
    Newline,
    CloseBrace,
    Newline,
    Newline,
    Comment(
        "Setup the graphics stack for bios and efi systems",
    ),
    If,
    Value(
        "[",
    ),
    Value(
        "${grub_platform}",
    ),
    Value(
        "=",
    ),
    Value(
        "efi",
    ),
    CloseBracket,
    Semicolon,
    Then,
    Newline,
    Value(
        "insmod",
    ),
    Value(
        "efi_gop",
    ),
    Newline,
    Value(
        "insmod",
    ),
    Value(
        "efi_uga",
    ),
    Newline,
    Else,
    Newline,
    Value(
        "insmod",
    ),
    Value(
        "vbe",
    ),
    Newline,
    Fi,
    Newline,
    Newline,
    Value(
        "insmod",
    ),
    Value(
        "font",
    ),
    Newline,
    If,
    Value(
        "loadfont",
    ),
    Value(
        "($drive1)//converted-font.pf2",
    ),
    Semicolon,
    Then,
    Newline,
    Value(
        "insmod",
    ),
    Value(
        "gfxterm",
    ),
    Newline,
    If,
    Value(
        "[",
    ),
    Value(
        "${grub_platform}",
    ),
    Value(
        "=",
    ),
    Value(
        "efi",
    ),
    CloseBracket,
    Semicolon,
    Then,
    Newline,
    Value(
        "set",
    ),
    Value(
        "gfxmode=auto",
    ),
    Newline,
    Value(
        "set",
    ),
    Value(
        "gfxpayload=keep",
    ),
    Newline,
    Else,
    Newline,
    Value(
        "set",
    ),
    Value(
        "gfxmode=1024x768",
    ),
    Newline,
    Value(
        "set",
    ),
    Value(
        "gfxpayload=text",
    ),
    Newline,
    Fi,
    Newline,
    Value(
        "terminal_output",
    ),
    Value(
        "gfxterm",
    ),
    Newline,
    Fi,
    Newline,
    Newline,
    Value(
        "background_color",
    ),
    Literal(
        "#2F302F",
    ),
    Newline,
    Newline,
    Value(
        "insmod",
    ),
    Value(
        "png",
    ),
    Newline,
    If,
    Value(
        "background_image",
    ),
    Value(
        "--mode",
    ),
    Literal(
        "normal",
    ),
    Value(
        "($drive1)//background.png",
    ),
    Semicolon,
    Then,
    Newline,
    Value(
        "set",
    ),
    Value(
        "color_normal=white/black",
    ),
    Newline,
    Value(
        "set",
    ),
    Value(
        "color_highlight=black/white",
    ),
    Newline,
    Else,
    Newline,
    Value(
        "set",
    ),
    Value(
        "menu_color_normal=cyan/blue",
    ),
    Newline,
    Value(
        "set",
    ),
    Value(
        "menu_color_highlight=white/blue",
    ),
    Newline,
    Fi,
    Newline,
    Newline,
    Newline,
    Value(
        "menuentry",
    ),
    Value(
        "NixOS - Default",
    ),
    Value(
        "--class",
    ),
    Value(
        "nixos",
    ),
    Value(
        "--unrestricted",
    ),
    OpenBrace,
    Newline,
    Value(
        "search",
    ),
    Value(
        "--set=drive1",
    ),
    Value(
        "--fs-uuid",
    ),
    Value(
        "BB22-99EC",
    ),
    Newline,
    Value(
        "linux",
    ),
    Value(
        "($drive1)//kernels/1pzgainlvg5hcdf8ngjficg3x39j63gv-linux-6.0.15-bzImage",
    ),
    Value(
        "init=/nix/store/kd6rrm02cs95hd8mf1kmnx14vksnrhwi-nixos-system-beetroot-23.05.20230102.0fc9fca/init",
    ),
    Value(
        "loglevel=4",
    ),
    Value(
        "zram.num_devices=1",
    ),
    Newline,
    Value(
        "initrd",
    ),
    Value(
        "($drive1)//kernels/2spacgl5d8l701wskfrz45q575f0q3fg-initrd-linux-6.0.15-initrd",
    ),
    Newline,
    CloseBrace,
    Newline,
    Newline,
    Newline,
    Value(
        "menuentry",
    ),
    Value(
        "NixOS - (flashfriendly - 1969-12-31 - 23.05.20230102.0fc9fca)",
    ),
    OpenBrace,
    Newline,
    Value(
        "search",
    ),
    Value(
        "--set=drive1",
    ),
    Value(
        "--fs-uuid",
    ),
    Value(
        "BB22-99EC",
    ),
    Newline,
    Value(
        "linux",
    ),
    Value(
        "($drive1)//kernels/1pzgainlvg5hcdf8ngjficg3x39j63gv-linux-6.0.15-bzImage",
    ),
    Value(
        "init=/nix/store/0y1mijr8m7wdah8mzlc31g5gj8w7psal-nixos-system-beetroot-23.05.20230102.0fc9fca/init",
    ),
    Value(
        "iomem=relaxed",
    ),
    Value(
        "loglevel=4",
    ),
    Value(
        "zram.num_devices=1",
    ),
    Newline,
    Value(
        "initrd",
    ),
    Value(
        "($drive1)//kernels/2spacgl5d8l701wskfrz45q575f0q3fg-initrd-linux-6.0.15-initrd",
    ),
    Newline,
    CloseBrace,
    Newline,
    Newline,
    Value(
        "submenu",
    ),
    Value(
        "NixOS - All configurations",
    ),
    Value(
        "--class",
    ),
    Value(
        "submenu",
    ),
    OpenBrace,
    Newline,
    Value(
        "menuentry",
    ),
    Value(
        "NixOS - Configuration 78 (2023-01-04 - 23.05.20230102.0fc9fca)",
    ),
    Value(
        "--class",
    ),
    Value(
        "nixos",
    ),
    OpenBrace,
    Newline,
    Value(
        "search",
    ),
    Value(
        "--set=drive1",
    ),
    Value(
        "--fs-uuid",
    ),
    Value(
        "BB22-99EC",
    ),
    Newline,
    Value(
        "linux",
    ),
    Value(
        "($drive1)//kernels/0wcixirfmmv0arr098ap4ynzk48npbmf-linux-6.1.1-bzImage",
    ),
    Value(
        "init=/nix/store/sj68yqfhzfq0h3bgi2v6dj464ak2v7bb-nixos-system-okra-23.05.20230102.0fc9fca/init",
    ),
    Value(
        "loglevel=4",
    ),
    Value(
        "zram.num_devices=1",
    ),
    Newline,
    Value(
        "initrd",
    ),
    Value(
        "($drive1)//kernels/vc142a0m33nvrcy44lrz5xph3n7gr756-initrd-linux-6.1.1-initrd",
    ),
    Newline,
    CloseBrace,
    Newline,
    Newline,
    Value(
        "menuentry",
    ),
    Value(
        "NixOS - Configuration 77 (2023-01-03 - 23.05.20230102.0fc9fca)",
    ),
    Value(
        "--class",
    ),
    Value(
        "nixos",
    ),
    OpenBrace,
    Newline,
    Value(
        "search",
    ),
    Value(
        "--set=drive1",
    ),
    Value(
        "--fs-uuid",
    ),
    Value(
        "BB22-99EC",
    ),
    Newline,
    Value(
        "linux",
    ),
    Value(
        "($drive1)//kernels/1pzgainlvg5hcdf8ngjficg3x39j63gv-linux-6.0.15-bzImage",
    ),
    Value(
        "init=/nix/store/q42sjmlgdzhr7zyq7l3p41lhbzldwz09-nixos-system-okra-23.05.20230102.0fc9fca/init",
    ),
    Value(
        "loglevel=4",
    ),
    Value(
        "zram.num_devices=1",
    ),
    Newline,
    Value(
        "initrd",
    ),
    Value(
        "($drive1)//kernels/vkvh2cjaybp0gf18k1v8pxzgwd3nch3v-initrd-linux-6.0.15-initrd",
    ),
    Newline,
    CloseBrace,
    Newline,
    Newline,
    CloseBrace,
    Newline,
]
